examples/toy_mixture_model.py

```diff
@@ -0,0 +1,163 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+ A toy mixture model to provide a simple example for implementing discrete enumeration.
+
+ (A) -> [B] -> (C)
+
+ A is an observed Bernoulli variable with Beta prior.
+ B is a hidden variable which is a mixture of two Bernoulli distributions (with Beta priors),
+ chosen by A being true or false.
+ C is observed, and like B, is a mixture of two Bernoulli distributions (with Beta priors),
+ chosen by B being true or false.
+ There is a plate over the three variables for n independent observations of data.
+
+ Because B is hidden and discrete we wish to marginalize it out of the model.
+ This is done by:
+    1) marking the model method with `@pyro.infer.config_enumerate`
+    2) marking the B sample site in the model with `infer={"enumerate": "parallel"}`
+    3) passing `pyro.infer.SVI` the `pyro.infer.TraceEnum_ELBO` loss function
+"""
+
+import os
+
+import matplotlib.pyplot as plt
+import numpy as np
+import torch
+from torch.distributions import constraints
+from torch.distributions.bernoulli import Bernoulli
+from torch.distributions.beta import Beta
+from tqdm import tqdm
+
+import pyro
+import pyro.distributions as dist
+import pyro.infer
+import pyro.optim
+from pyro.ops.indexing import Vindex
+
+smoke_test = ('CI' in os.environ)
+
+
+def main():
+
+    # number of observations
+    n = 10000
+
+    prior, CPDs, data = generate_data(n)
+
+    posterior_params = train(prior, data, n)
+
+    evaluate(CPDs, posterior_params)
+
+
+def generate_data(n):
+
+    # domain = [False, True]
+    prior = {'A': torch.tensor([1., 10.]),
+             'B': torch.tensor([[10., 1.],
+                                [1., 10.]]),
+             'C': torch.tensor([[10., 1.],
+                                [1., 10.]]),
+             }
+
+    # CPDs
+    CPDs = {'p_A': Beta(prior['A'][0], prior['A'][1]).sample(),
+            'p_B': Beta(prior['B'][:, 0], prior['B'][:, 1]).sample(),
+            'p_C': Beta(prior['C'][:, 0], prior['C'][:, 1]).sample(),
+            }
+
+    data = {}
+    data['A'] = Bernoulli(torch.ones(n) * CPDs['p_A']).sample()
+    data['B'] = Bernoulli(torch.gather(CPDs['p_B'], 0, data['A'].type(torch.long))).sample()
+    data['C'] = Bernoulli(torch.gather(CPDs['p_C'], 0, data['B'].type(torch.long))).sample()
+
+    return prior, CPDs, data
+
+
+@pyro.infer.config_enumerate
+def model(prior, obs, n):
+
+    p_A = pyro.sample('p_A', dist.Beta(1, 1))
+
+    p_B = pyro.sample('p_B', dist.Beta(torch.ones(2), torch.ones(2)).to_event(1))
+
+    p_C = pyro.sample('p_C', dist.Beta(torch.ones(2), torch.ones(2)).to_event(1))
+
+    with pyro.plate('data_plate', n):
+        A = pyro.sample('A', dist.Bernoulli(p_A.expand(n)), obs=obs['A'])
+
+        B = pyro.sample('B', dist.Bernoulli(Vindex(p_B)[A.type(torch.long)]), infer={"enumerate": "parallel"})
```

Comment:
could you maybe add a comment on the role of `Vindex`?

Reply:
Done.